### PR TITLE
Support for OSTREE-based distributions

### PR DESCRIPTION
**Pull request by PilotGuy772**
Fixes https://github.com/bayasdev/envycontrol/issues/180
Fixes https://github.com/bayasdev/envycontrol/issues/114
Fixes https://github.com/bayasdev/envycontrol/issues/49

EnvyControl is configured right now to put udev rules in `/lib/udev`. The problem with this is that `/lib/udev` is distinctly intended for udev rules from the distribution. It is NOT meant to be manipulated so heavily. Instead, system admins ought to use `/etc/udev` to put custom rules. 

This ordinarily isn't a problem for EnvyControl because most distros don't care if you modify `/lib/`, but immutable distros like Fedora Silverblue *do* care. `/lib/` is not writable, but `/etc/` is. There's a reason for that.

This PR **changes the udev directory to `/etc/udev/`**, but everything still works as expected with this change. I also added logic to check if the distribution uses ostree by checking for `/ostree` and `/sysroot/ostree`, which are the two places where you would usually find the ostree directory. My amateur testing on Bluefin has shown that switching between integrated and hybrid modes works perfectly.

### CODE REVIEW

File: envycontrol.py [hunk 1]
```diff
@@ -38,7 +38,7 @@
 alias nvidia_current_modeset off
 '''
 
-UDEV_INTEGRATED_PATH = '/lib/udev/rules.d/50-remove-nvidia.rules'
+UDEV_INTEGRATED_PATH = '/etc/udev/rules.d/50-remove-nvidia.rules'
```
Comment (bayasdev) [inline envycontrol.py:41]: @PilotGuy772 could you please add a new variable instead? As systems running older versions may have these files on their previous location the cleanup routine will not remove them

File: envycontrol.py [hunk 2]
```diff
@@ -55,7 +55,7 @@
 ACTION=="add", SUBSYSTEM=="pci", ATTR{vendor}=="0x10de", ATTR{class}=="0x03[0-9]*", ATTR{power/control}="auto", ATTR{remove}="1"
 '''
 
-UDEV_PM_PATH = '/lib/udev/rules.d/80-nvidia-pm.rules'
+UDEV_PM_PATH = '/etc/udev/rules.d/80-nvidia-pm.rules'
```
Comment (bayasdev) [inline envycontrol.py:58]: Same here

**Review by bayasdev — CHANGES_REQUESTED** [review]

File: envycontrol.py (hunk 1 above)
Comment (bayasdev) [reply envycontrol.py:41]: Another solution would be to manually add the paths to the cleanup routine that way we won't need a new variable

**Comment by PilotGuy772** [comment]
@bayasdev I added logic to cleanup() in my latest commit to address your concerns. I just have statements to check for EnvyControl's udev rules in `/lib/udev/rules.d` and remove them if they exist.

File: envycontrol.py [hunk 3]
```diff
@@ -367,6 +367,23 @@ def cleanup():
             if e.errno != 2:
                 logging.error(f"Failed to remove file '{file_path}': {e}")
 
+    # attempt to delete rules from legacy directory
+    legacy_udev = '/lib/udev/rules.d/50-remove-nvidia.rules'
+    legacy_udev_pm = '/lib/udev/rules.d/80-nvidia-pm.rules'
+    if os.path.exists(legacy_udev):
```
Comment (bayasdev) [inline envycontrol.py:373]: @PilotGuy772 I think we could do something like this instead

```
    # define list of files to remove
    to_remove = [
        BLACKLIST_PATH,
        UDEV_INTEGRATED_PATH,
        UDEV_PM_PATH,
        XORG_PATH,
        EXTRA_XORG_PATH,
        MODESET_PATH,
        LIGHTDM_SCRIPT_PATH,
        LIGHTDM_CONFIG_PATH,
        # legacy paths
        '/etc/X11/xorg.conf.d/90-nvidia.conf',
        '/lib/udev/rules.d/50-remove-nvidia.rules',
        '/lib/udev/rules.d/80-nvidia-pm.rules'
    ]

    # remove each file in the list
    for file_path in to_remove:
        try:
            if os.path.exists(lfile_path):
                os.remove(file_path)
                logging.info(f"Removed file {file_path}")
        except OSError as e:
            # only warn if file exists (code 2)
            if e.errno != 2:
                logging.error(f"Failed to remove file '{file_path}': {e}")
```

File: envycontrol.py [hunk 4]
```diff
@@ -463,8 +480,23 @@ def get_amd_igpu_name():
...
+    if any(os.path.exists(dir) for dir in ['/ostree', '/sysroot/ostree']):
+        # As per the rpm-ostree manpage, the command rpm-ostree initramfs
+        # may be used to regenerate the iintramfs with dracut. This will
+        # run a full regeneration, similar to what is used in the other
+        # trees of this if/elif
+
+        # I'm keeping --force, but --regenerate-all is not necessary
+        # because rpm-ostree already passes the kernel version.
+
+        # this takes a LLOOOOOONG time to run, so let's warn the user first.
```
Comment (bayasdev) [inline envycontrol.py:494]: could you please remove the redundant comments?

File: envycontrol.py [hunk 5]
```diff
@@ -600,7 +632,7 @@ class CachedConfig:
 
     def __init__(self, app_args) -> None:
         self.app_args = app_args
-        self.current_mode = get_current_mode()
+        self.current_mode = get_nvidia_gpu_pci_bus()
```
Comment (bayasdev) [inline envycontrol.py:635]: why was this changed?

**Review by bayasdev — CHANGES_REQUESTED** [review]

**Comment by bayasdev** [comment]
@PilotGuy772 please take a look at the comments I left, and thank you for adding OSTree support to EnvyControl as it was a frequently requested feature 🙌🏼

**Comment by PilotGuy772** [comment]
@bayasdev All done (I think)

**Review by bayasdev — APPROVED** [review]
LGTM

**Comment by bayasdev** [comment]
Thanks @PilotGuy772

**Pull request merged**